services/dedup/scylla/scylla.go

```diff
@@ -0,0 +1,135 @@
+package scylla
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/gocql/gocql"
+	"github.com/samber/lo"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/stats"
+	"github.com/rudderlabs/rudder-server/services/dedup/types"
+)
+
+type ScyllaDB struct {
+	scylla *gocql.Session
+	conf   *config.Config
+	stat   stats.Stats
+	// Time to live in seconds for an entry in the DB
+	ttl int
+	// Maximum number of keys to commit in a single batch
+	batchSize      int
+	createTableMu  sync.Mutex
+	createTableMap map[string]*sync.Once
+
+	cacheMu  sync.Mutex
+	cache    map[string]types.KeyValue
+	keyspace string
+}
+
+func (d *ScyllaDB) Close() {
+	d.scylla.Close()
+}
+
+func (d *ScyllaDB) Get(kv types.KeyValue) (bool, int64, error) {
+	// Create the table if it doesn't exist
+	var err error
+	d.createTableMu.Lock()
+	once, found := d.createTableMap[kv.WorkspaceId]
+	if !found {
+		d.createTableMap[kv.WorkspaceId] = &sync.Once{}
+		once = d.createTableMap[kv.WorkspaceId]
+	}
+	d.createTableMu.Unlock()
+	once.Do(func() {
+		query := fmt.Sprintf("CREATE TABLE IF NOT EXISTS %s.%s (id text PRIMARY KEY,size bigint) WITH bloom_filter_fp_chance = 0.005;", d.keyspace, kv.WorkspaceId)
+		err = d.scylla.Query(query).Exec()
+	})
+	if err != nil {
+		return false, 0, fmt.Errorf("error creating table %s: %v", kv.WorkspaceId, err)
+	}
+
+	d.cacheMu.Lock()
+	defer d.cacheMu.Unlock()
+	// Check if the key exists in the cache
+	// This is essential if we get the same key multiple times in the same batch
+	// Since we are not committing the keys immediately, we need to keep track of the keys in the cache
+	if previous, found := d.cache[kv.Key]; found {
+		return false, previous.Value, nil
+	}
+
+	// Check if the key exists in the DB
+	var value int64
+	err = d.scylla.Query(fmt.Sprintf("SELECT size FROM %s.%s WHERE id = ?", d.keyspace, kv.WorkspaceId), kv.Key).Scan(&value)
+	if err != nil && err != gocql.ErrNotFound {
```

Comment:
Are you positive this never gets wrapped? Would it make sense to go with `errors.Is` just in case?

Reply:
Updated